scoped_model_counter/test/widget_test.dart

```diff
@@ -0,0 +1,28 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:scoped_model/scoped_model.dart';
+import 'package:scoped_model_counter/main.dart';
+
+void main() {
+  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
+    // Build our app, provide it with a model, and trigger a frame.
+    await tester.pumpWidget(
+      ScopedModel(
+        model: Counter(),
+        child: MyApp(),
+      ),
+    );
+
+    // Verify that our counter starts at 0.
+    expect(find.text('0'), findsOneWidget);
+    expect(find.text('1'), findsNothing);
+
+    // Tap the '+' icon and trigger a frame.
+    await tester.tap(find.byIcon(Icons.add));
+    await tester.pump();
+
+    // Verify that our counter has incremented.
+    expect(find.text('0'), findsNothing);
+    expect(find.text('1'), findsOneWidget);
+  });
```

Comment:
Given the speed of tests, the 5 sec outside interaction hasn't happened yet.

Maybe we could add a delay here and check that the counter gets updated from out-of-band?

Reply:
+1 to this, mainly so I can see another example of how to do async properly in tests. 😄

Reply:
The outside interaction isn't tested here (it's not set up, we're just pumping the site).

The reason for that is that I would have to extract the outside interaction to something that could be added here. Which adds complexity to an otherwise simple example. 

If I don't extract, then I need to duplicate the timer code inside the test, which defeats the purpose of the test.

Happy to add an async sample to some other test in this repo.

Reply:
> The outside interaction isn't tested here (it's not set up, we're just pumping the site).

Ah, right! :( ... hmm. 

It might not be much more complex to factor out the model creation and timer setup code into a function, maybe something like this:

```dart
Widget MyAppWithScoppedModel() {
  // Initialize the model. Can be done outside a widget, like here.
  var counter = Counter();

  // Setup a delayed interaction with the model (increment each 5 seconds),
  // outside of the Flutter widget tree.
  Timer.periodic(
    const Duration(seconds: 5),
    (timer) => counter.increment(),
  );

  // Provide the model to all the app's widgets
  return ScopedModel<Counter>(
    model: counter,
    child: MyApp(),
  );
}

// Run the app using a scoped model:
void main() => runApp(MyAppWithScoppedModel());
```

Then you can `pumpWidget(MyAppWithScoppedModel())`. WDYT?

Reply:
Now that I've had a chance to think more about this, I'm inclined toward Filip's "let's just keep it simple" approach. IMHO, What's being tested here is:

* Whether the app correctly builds and rebuilds widgets based on a ScopedModel.
* Whether the app can correctly update a ScopedModel instance based on a UI event.

I think that covers what's needed. Adding the Timer into the test would get you:

* Whether the Timer code in the app is correctly updating the ScopedModel.
* Whether ScopedModel is correctly notifying its listeners when updated asynchronously by a Timer.

The first of those is within the domain of the app, but not all that interesting. The second is more of a ScopedModel concern than something to specific to this app.

Reply:
👍